### ISSUE TITLE
"mob timer" doesn't do anything in the terminal

### ISSUE THREAD

**Issue by schmonz** [remotemobprogramming/mob#158]
I'm trying `mob` in a use case that's probably not common: sharing a remote terminal (via [tmate](https://tmate.io)), not needing the WIP branch management stuff, just looking for a timer that can notify us inside the terminal.

I tried running `mob timer 1`, hoping that after a minute the terminal bell would ring (or perhaps flash). As far as I can tell, nothing happened. 

I tried running the same command locally and immediately got a "mob next" notification. I guess that's probably too fast to be the currently intended behavior. (Both test systems were macOS Big Sur, in case that matters.)

Would it be possible for `mob timer` to ring the terminal bell?

**Comment by simonharrer** [comment]
what is a terminal bell?

**Comment by simonharrer** [comment]
if there’s a way to trigger that what you want you can configure that via an environment variable. key in mob config to see the commands for the notification. You set the variable to the command that should be called.

**Comment by simonharrer** [comment]
For example, you could use 

```bash
MOB_NOTIFY_COMMAND="/usr/bin/osascript -e 'beep'" mob t 1
```

If you also want to disable the voice command, you can use:

```bash
MOB_NOTIFY_COMMAND="/usr/bin/osascript -e 'beep'" MOB_VOICE_COMMAND= mob t 1
```

**Comment by schmonz** [comment]
Thanks for the tip. I'm getting somewhere.

From [Rosetta Code](https://rosettacode.org/wiki/Terminal_control/Ringing_the_terminal_bell),

> In most terminals, if the [Bell character](https://en.wikipedia.org/wiki/Bell_character) (ASCII code **7**, `\a` in C)   is printed by the program, it will cause the terminal to ring its bell.   This is a function of the terminal, and is independent of the programming language of the program, other than the ability to print a particular character to standard out.

If I `export MOB_NOTIFY_COMMAND='echo ^G^G^G'` (where each `^G` is the result of typing `Ctrl-V` followed by `Ctrl-G`, so as to write a `Ctrl-G` character into the string instead of beeping right now), then I think `mob t 1` almost does what I want. The remaining difficulty is that `mob t 1` immediately mutes the macOS system audio output, whether or not I override `MOB_VOICE_COMMAND` with an empty string.

**Comment by simonharrer** [comment]
You can debug what mob is doing with the `--debug` flag. It doesn't do a lot of magic, actually. 

Here's the output for me:
```
➜  mob git:(main) MOB_NOTIFY_COMMAND='echo ^G^G^G' MOB_VOICE_COMMAND= mob t 0 --debug
 DEBUG overriding MOB_VOICE_COMMAND=
 DEBUG overriding MOB_NOTIFY_COMMAND=echo ^G^G^G
 DEBUG found MOB_NEXT_STAY=true
 DEBUG overriding MOB_NEXT_STAY =true
 DEBUG Args 'mob t 0 --debug'
 DEBUG command 't'
 DEBUG parameters '0'
 DEBUG version 1.3.0
 DEBUG workingDir
 DEBUG Starting timer at 13:32 for 0 minutes = 0 seconds (parsed from user input 0)
 DEBUG Operating System darwin
 DEBUG Starting command sh -c (sleep 0;echo ^G^G^G mob next) &
 > 0 minutes timer started (finishes at approx. 13:32)
```

Relevant is probably only ` sh -c (sleep 0;echo ^G^G^G mob next) &`, but bear in mind that quotes might be missing in this console output. This basically executes a shell expression in the background. Perhaps this helps you.

**Comment by simonharrer** [comment]
Anything I can do to help you?

**Comment by schmonz** [comment]
Thanks for the ping. I think I'm having a few overlapping difficulties:

1. `mob timer 1000` gives a "mob next" notification immediately; I (perhaps incorrectly?) expected to see it only after 1000 minutes
2. When I see "mob next", I don't hear it; `mob timer` and `mob start` immediately mute the macOS system and alert volumes (whether with custom settings or the defaults)
3. Once my volume can stay where I've left it, I can figure out whether `MOB_NOTIFY_COMMAND='echo ^G^G^G'` is the config I'm looking for

Is (1) a bug? Any guesses why (2) might be happening? FWIW, `mob moo` does _not_ change the volume settings and I hear it just fine.

**Comment by simonharrer** [comment]
mob timer only executes a command on the console and fork it so it runs in a separate process. It sleeps the amount of seconds until the timer is up and then runs commands.

Can you provide the —debug output for mob timer 1000, please?

And I have no idea why mob would mute anything in your system. I cannot think of any code in the mob tool that could trigger such behavior.

**Comment by schmonz** [comment]
```
$ mob timer 1000 --debug
 DEBUG key MOB_NEXT_STAY is not set
 DEBUG Args 'mob timer 1000 --debug'
 DEBUG command 'timer'
 DEBUG parameters '1000'
 DEBUG version 1.4.0
 DEBUG workingDir
 DEBUG Starting timer at 10:40 for 1000 minutes = 60000 seconds (parsed from user input 1000)
 DEBUG Operating System darwin
 DEBUG Starting command sh -c (sleep 60000;say "mob next";/usr/bin/osascript -e 'display notification "mob next"') &
 > 1000 minutes timer started (finishes at approx. 10:40)
$
```

And with `time` added to the command line:

```
$ time mob timer 1000 --debug
 DEBUG key MOB_NEXT_STAY is not set
 DEBUG Args 'mob timer 1000 --debug'
 DEBUG command 'timer'
 DEBUG parameters '1000'
 DEBUG version 1.4.0
 DEBUG workingDir
 DEBUG Starting timer at 10:43 for 1000 minutes = 60000 seconds (parsed from user input 1000)
 DEBUG Operating System darwin
 DEBUG Starting command sh -c (sleep 60000;say "mob next";/usr/bin/osascript -e 'display notification "mob next"') &
 > 1000 minutes timer started (finishes at approx. 10:43)

real	0m0.268s
user	0m0.003s
sys	0m0.010s
$
```

I do have some audio extensions installed. It's possible there's a bad interaction between some of them, or maybe between one of them and the go runtime. I'll try rebooting in "Safe Mode" (with the Shift key held down) and see if that gets rid of the muting and/or somehow corrects the timer behavior.

**Comment by schmonz** [comment]
Safe Mode turns off all the regular Mac audio devices, too :-/  Putting the strange muting problem on hold, though, because I've figured out when the background command doesn't seem to work and when it does.

### Doesn't work

```
$ MOB_NOTIFY_COMMAND="echo 'SCHMONZ: mob next' | write $(whoami)" mob timer 0 --debug
 DEBUG overriding MOB_VOICE_COMMAND=
 DEBUG overriding MOB_NOTIFY_COMMAND=echo 'SCHMONZ: mob next' | write classroom
 DEBUG key MOB_NEXT_STAY is not set
 DEBUG Args 'mob timer 0 --debug'
 DEBUG command 'timer'
 DEBUG parameters '0'
 DEBUG version 1.4.0
 DEBUG workingDir
 DEBUG Starting timer at 05:33 for 0 minutes = 0 seconds (parsed from user input 0)
 DEBUG Operating System darwin
 DEBUG Starting command sh -c (sleep 0;echo 'SCHMONZ: mob next' | write classroom mob next) &
 > 0 minutes timer started (finishes at approx. 05:33)
$
```

### Does work

```
$ MOB_NOTIFY_COMMAND="echo 'SCHMONZ: mob next' | write $(whoami);" mob timer 0 --debug
 DEBUG overriding MOB_VOICE_COMMAND=
 DEBUG overriding MOB_NOTIFY_COMMAND=echo 'SCHMONZ: mob next' | write classroom;
 DEBUG key MOB_NEXT_STAY is not set
 DEBUG Args 'mob timer 0 --debug'
 DEBUG command 'timer'
 DEBUG parameters '0'
 DEBUG version 1.4.0
 DEBUG workingDir
 DEBUG Starting timer at 05:32 for 0 minutes = 0 seconds (parsed from user input 0)
 DEBUG Operating System darwin
 DEBUG Starting command sh -c (sleep 0;echo 'SCHMONZ: mob next' | write classroom; mob next) &
 > 0 minutes timer started (finishes at approx. 05:32)

Message from classroom@pet-power-plant.local on tty?? at 05:32 ...
SCHMONZ: mob next
EOF
$
```

(And if I give a non-zero number of minutes, that works too.)

The difference-maker is whether my `MOB_NOTIFY_COMMAND` definition includes a trailing `;` of its own. If it does, then the background command properly delimits my command from `mob next` and I see the result of my command; if it doesn't, they're not delimited as they should be (and `mob next` are probably getting passed as two extra arguments to my command).

So now I can say something more precise about how the current behavior is unexpected, for me: I expected `mob` to handle delimiting my custom command from any internal commands it may also run. I guess this is a bug somewhere around line 442 (in 1.4, anyway), in `executeCommandsInBackgroundProcess()`.

**Comment by simonharrer** [comment]
The MOB_NOTIFY_COMMAND either has a `%s` which is substituted by `mob next` or, in case it has not, `mob next` is appended to the MOB_NOTIFY_COMMAND string before processed further. I think that's the issue here.

**Comment by schmonz** [comment]
OMG, the problem with muting was massive and hilarious PEBKAC. I use [SleepWatcher](https://www.bernhard-baehr.de) to mute my volume when closing the lid, so that when I open it it's quiet by default. My little mute script was called `sleep` and it was in my `PATH`. I figured this out by trying `sleep 60` directly, seeing it return immediately, and getting a surprising result for `which sleep`. My script is called something else now.

`mob moo` doesn't call `sleep` at all, of course, and that's why I could hear it just fine.

In conclusion, somewhat predictably, 🤦🏼‍♂️. There's no problem with `mob` and the system's audio. Only with the way the custom command line is assembled.

**Comment by simonharrer** [comment]
Okay, didn't expect this. Glad this mystery is now solved! :-)

**Comment by simonharrer** [comment]
Closing this as it seems to be solved.

**Comment by schmonz** [comment]
Well, hang on. The audio problem was entirely on my end, indeed nothing for you to do about that. But the surprising behavior of `MOB_NOTIFY_COMMAND` -- requiring the user to semicolon-terminate its value -- looks to me like a bug in `mob`. I doubt I'm the only user who would never have guessed it was necessary to do that.

**Comment by simonharrer** [comment]
Instead of 

```
MOB_NOTIFY_COMMAND=echo 'SCHMONZ: mob next' | write classroom
```

can you try out 

```
MOB_NOTIFY_COMMAND=echo 'SCHMONZ: %s' | write classroom
```

I think that should do the trick.

**Comment by schmonz** [comment]
Ohhhhhhhhh! (It does.)

Okay, then maybe the essential detail I didn't realize is that custom non-empty values of `MOB_NOTIFY_COMMAND` and `MOB_VOICE_COMMAND` must not omit `%s`.

**Comment by simonharrer** [comment]
I agree that this is might be confusing. Not sure, however, how to handle this really well. One could require the notify command to contain a `%s` always. But perhaps not everybody wants this...

**Comment by schmonz** [comment]
What to do about this isn't clear to me either. I guess for now this issue documents one schmonz's encounter with the problem, and if and when it comes up for other people in other ways, maybe it'll become clear what's best to do.
